fix: incorrect templating not producing error

mikeykhalil opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

Fixes issue https://github.com/tektoncd/pipeline/issues/627.

This PR allows tasks to be more properly validated when being created. This allows users to catch templating errors earlier on. 

This PR also addresses an issue where resource variable validation would basically maintain a global list of variables. However, I believe we really want to create 2 list of variables. One list for input resource variables and one list for output resource variables. 

I am still not 100% convinced this is the best way to solve the issue, but it does seem to do the trick. Will sleep on it, but happy to address feedback on this PR or entertain other ideas on how to resolve this issue. 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md#principles) (if user facing) (note: I didn't write any docs for this, seems unnecessary).
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md)
for more details._

# Release Notes

<!--
Does your PR contain User facing changes?

If so, briefly describe them here so we can include this description in the
release notes for the next release!
-->
```
Return error to user for incorrect templating of resource variables.
```

mikeykhalil commented on pkg/apis/pipeline/v1alpha1/task_validation_test.go:
```diff
@@ -105,7 +105,7 @@ func TestTaskSpec_Validate(t *testing.T) {
 				Image:      "${inputs.resources.foo.url}",
 				Args:       []string{"--flag=${inputs.params.baz} && ${input.params.foo-is-baz}"},
-				WorkingDir: "/foo/bar/${outputs.resources.source}",
+				WorkingDir: "/foo/bar/${outputs.resources.source.name}",
```
This was actually an example of https://github.com/tektoncd/pipeline/issues/627. I had to modify this test in order for it to pass with the fix from this PR.

sbwsg commented on pkg/apis/pipeline/v1alpha1/task_validation.go:
```diff
@@ -132,67 +132,110 @@ func validateInputParameterVariables(steps []corev1.Container, inputs *Inputs) *
+	return validateVariables(steps, "resources", "outputs.", outputVars)
+}
+
+func getInputResourceVariables(inputs *Inputs) (map[string]struct{}, *apis.FieldError) {
```
I'm curious why you chose to use two separate functions for input & output here - it seems like it would be possible to write a single function to handle both with a signature like:

```go
func getResourceVariables(resources []TaskResource, pathPrefix string)
```

I might be misreading but it looks like the only line of difference between `getInputResourceVariables` and `getOutputResourceVariables` is the path string in the returned `apis.FieldError`, which could be resolved using a `pathPrefix` arg.

mikeykhalil replied:
You're totally right! There was plenty of duplicate code there. I went ahead and made these changes. Thanks for the suggestion.

sbwsg commented on pkg/templating/templating.go:
```diff
@@ -47,10 +47,7 @@ func extractVariablesFromString(s, prefix string) ([]string, bool) {
-		// foo.bar -> foo
-		// foo.bar.baz -> foo
-		vars[i] = strings.SplitN(groups["var"], ".", 2)[0]
+		vars[i] = groups["var"]
```
I'm having a hard time figuring out if the logic has remained the same here or has been changed to make an intentional difference to behaviour. Would you mind just quickly describing what this change is?

mikeykhalil replied:
Yes, I intentionally made this change. 

I added a new test to the templating package which hopefully provides additional insight as to why this was needed. This new test would fail without this change. 

Basically, `strings.SplitN(groups["var"], ".", 2)[0]` was just returning the first split of a variable. So if groups["var"] was set to `foo.bar.baz`, it would just return `foo`.

Since the `vars` map passed into `ValidateVariable` can contain nested variables now, like `fooImage.url`, I believe we want to return the entire variable string to validate against.

sbwsg replied:
OK got it, thankyou, I understand the original issue and this fix more clearly now!

sbwsg commented:
I pulled down the branch and tried out the validations by breaking and un-breaking a variable in the resource list of an example Task.  Worked well!

bobcatfish commented:
/ok-to-test

bobcatfish commented on pkg/apis/pipeline/v1alpha1/task_validation.go:
```diff
@@ -132,67 +132,110 @@ func validateInputParameterVariables(steps []corev1.Container, inputs *Inputs) *
+	return vars, nil
+}
+
+func getOutputResourceVariables(outputs *Outputs) (map[string]struct{}, *apis.FieldError) {
```
i think it's worth having unit tests for `getOutputResourceVariables` and `getInputResourceVariables` specifically - that might also help address @sbwsg 's feedback about knowing exactly what changed, b/c we'd be able to see it in the test cases

mikeykhalil replied:
This changed slightly and has been reduced to one function, `getResourceVariables`. Your comment is still completely valid though. I have an idea for a test that could be constructive for `getResourceVariables`, but I want to put a bit more thought into it so that it's not tightly coupled with the current structure of a given resource type. Will circle back to this tomorrow.

mikeykhalil commented:
Thanks for the initial reviews! Sorry for some of the sloppiness in the original PR. Ended up working on that past my bedtime and it showed 🤣. 

Anyways, I think I addressed most of the comments. 

2 items that still need to be addressed are:

1. Fix commit message to meet contributing standards. I was planning on squashing all of the commits and then writing a detailed commit message. Let me know if this is okay, otherwise, I can add details to each of the commit messages. 
2. Add tests for `getResourceVariables`.

bobcatfish commented on pkg/apis/pipeline/v1alpha1/resource_types.go:
```diff
@@ -164,3 +164,19 @@ func ResourceFromType(r *PipelineResource) (PipelineResourceInterface, error) {
+
+// AttributesFromType returns a list of available attributes that can be replaced
+// through templating.
+func AttributesFromType(prt PipelineResourceType) ([]string, error) {
```
we could have some unit tests for `AttributesFromType` too :D

mikeykhalil replied:
Glad you recommended this! Ended up running into an issue where this function was not working for one of the types (`PipelineResourceTypeStorage`). I included a fairly kludgy solution and a comment to help illustrate the problem.

> We have run into this unfortunate event where technically we cannot determine what attributes an object has strictly from the PipelineResourceType alone. This is because of the subtypes that exists for the PipelineResourceTypeStorage type. At the moment, both of the subtypes happen to have equivalent keys in their Replacements(). So technically we can just pick one of the subtypes and still return the proper attributes. However, we may want to reconsider the approach on how we obtain or store these attributes.

I do not have more time to look at this at the moment, but will find time soon to try to come up with a more elegant solution.

bobcatfish commented:
> Fix commit message to meet contributing standards. I was planning on squashing all of the commits and then writing a detailed commit message. Let me know if this is okay, otherwise, I can add details to each of the commit messages.

Sounds good!

> Add tests for getResourceVariables.

And `AttributesFromType` looks like another good candidate :D 

But generally I think this is nearly good to go @mikeykhalil , looking great! :D :tada: :tada: :tada:

vdemeester commented on pkg/apis/pipeline/v1alpha1/resource_types.go:
```diff
@@ -164,3 +164,39 @@ func ResourceFromType(r *PipelineResource) (PipelineResourceInterface, error) {
+
+			{
+				Name:  "type",
+				Value: string(PipelineResourceTypeGCS),
```
`PipelineResourceTypeStorage` could be something else than `PipelineResourceTypeGCS` though

mikeykhalil replied:
Correct, I think this is the blocking issue at the moment. I left a comment above in the code which goes into more detail. 

>We have run into this unfortunate event where technically we cannot determine what attributes an object has strictly from the PipelineResourceType alone. This is because of the subtypes that exists for the PipelineResourceTypeStorage type. At the moment, both of the subtypes happen to have equivalent keys in their Replacements(). So technically we can just pick one of the subtypes and still return the proper attributes. However, we may want to reconsider the approach on how we obtain or store these attributes.

Still not exactly sure how to go about solving this problem, but I will think about it some more and update this PR with a better solution once I find the time.

sbwsg replied:
It seems like there isn't enough information in a `TaskResource` struct to correctly infer the type of a `PipelineResourceTypeStorage`. The exact storage type only gets determined when a `TaskRun` is created I think, right?  It doesn't seem to me like there's a way to validate storage when the Task is applied since its type parameter wont be declared until later.

Options I can see right now are:
1. Leave the code as it is :- make an assumption about the storage type. Future storage types will need to include these keys to validate or will need to change this validation code.
2. Augment the `TaskResource` with enough info to infer the correct storage type to validate against. This seems counter to the design of PipelineResourceTypeStorage - the dynamic binding of the storage's exact type is intentionally left until the TaskRun to allow different sources to feed through a pipeline, I think?
3. Agree on a set of templatable params that all storage types must implement.
4. Iterate over all known storage types and check whether the templated fields match at least one of those storage kinds.
5. Abandon validating storage params in a Task template.  Maybe do it in TaskRuns instead.

Have I misunderstood the implementation of `TaskResource` / `TaskResourceBinding`?  Are there other options available to us here?

mikeykhalil replied:
Your analysis seems correct to me and I was thinking about a few of those options as well. 

I agree that option 2 seems to counter the current design, though it's possible I have a misunderstanding.

I think the rest of the options you mentioned are solid approaches.

In my mind, option 3 sounds like an ideal approach and a convenient characteristic for all types of `PipelineResourceTypeStorage` to share. However, this might not be realistic depending on the types of `PipelineResourceTypeStorage` we plan to support in the future.

Option 4 seems like something that could be built now with not much effort and provide immediate benefit. There could be some cases in the future where different 
 types of `PipelineResourceTypeStorage` don't share the same templatable params, in which case we would only be able to catch certain errors at `TaskRun` execution time.

Option 5 seems reasonable and provides similar benefits as option 4. We will catch most kinds of templating errors earlier on but still have certain errors at `TaskRun` execution time.

---

After some thought, if we feel we could reach a consensus on option 3 then that seems like the most robust approach. 

If that's not something we think we can enforce, then option 4 or 5 sound great to me.

sbwsg replied:
Option 3 seems like a design that would be worth raising with the community - I doubt we could confidently come to a widely agreeable consensus in this github thread alone.  So the process for that would be to create a github issue describing the problem along with a suggested resolution and design doc if the problem is gnarly enough (this one is, I think).

It may also be that we can push this specific problem back until there is more movement on [Pipeline Resource Extensibility](https://github.com/tektoncd/pipeline/issues/238), which may end up setting the "validatable surface area" simply as a byproduct of the interfaces that it defines for extensions.

Having said all that it seems to me that we could move ahead with options 4 or 5 in this PR and work on option 3 out-of-band.  If it were me I'd probably skip validating storage types completely, leave a comment explaining why, and create a github issue to describe the problem and possible solutions in detail.

sbwsg replied:
Actually thinking about this some more, I reckon simply pushing the validation to the TaskRun is totally fine.  The whole point of the `PipelineResourceTypeStorage` is that it's a placeholder to be filled in later. The Task shouldn't need to really worry about validating it deeply.  At least from my POV.

sbwsg replied:
I'm now realizing that option 5, skipping storage params, is kind of tricky. It's not possible to easily express "ignore variables of this type" when doing the template validation.

mikeykhalil replied:
Ah, if the task shouldn't have to worry about validating it deeply, then it's possible that we do not really want this kind of patch and may just want to revisit this in a more broad way with a github issue and a design doc at some point.

Otherwise, it might be possible to go with a hybrid approach where we deeply validate all types of variables besides storage. Then for storage params, we just validate the top level variable (similar to how the validation works currently). But yeah, not sure if there is an easy way to express that as you said.

Been busy so I haven't had time to take a look, but I can take a look this weekend to see if i can come up with a nice way to solve it.

dlorenc commented:
It looks like this needs a small rebase now, are you still interested in getting this finished up @mikeykhalil ?

mikeykhalil commented:
> It looks like this needs a small rebase now, are you still interested in getting this finished up @mikeykhalil ?

Hey, sorry for the lack of activity from me. A few things came up and I was out of town for a bit. 

It seems like there is not a clear way to deeply validate all kinds of variables except for those of type `storage`. And even if there was, the behavior would be inconsistent and possibly confusing for users. 

For now, I think the consensus is that we are better off closing this PR and maybe addressing this issue  once https://github.com/tektoncd/pipeline/issues/238 gains more traction. 

Thanks for all of the help, everyone! It's a bummer we couldn't get something merged to improve the validation at task creation time, but I think we at least have some more context for how we may want to address the issue in the future. Feel free to reopen or ping me if I can help in some way.

bobcatfish commented:
Thanks for all your hard work @mikeykhalil !!! Thanks for helping us investigate and understand this problem better.
